pkg/envoy/memoize/memoize.go

```diff
@@ -0,0 +1,53 @@
+package memoize
+
+import (
+	"time"
+
+	xds_discovery "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3"
+	"github.com/patrickmn/go-cache"
+
+	"github.com/openservicemesh/osm/pkg/catalog"
+	"github.com/openservicemesh/osm/pkg/certificate"
+	"github.com/openservicemesh/osm/pkg/configurator"
+	"github.com/openservicemesh/osm/pkg/envoy"
+	"github.com/openservicemesh/osm/pkg/logger"
+)
+
+var log = logger.New("envoy/memoize")
+
+const (
+	defaultExpiration = 1 * time.Second
+	cleanupInterval   = 10 * time.Minute
+)
+
+var memoizationCache = make(map[string]*cache.Cache)
+
+// Memoize memoizes the function given
+func Memoize(
+	dsType, key string,
+	fn func(catalog.MeshCataloger, *envoy.Proxy, *xds_discovery.DiscoveryRequest, configurator.Configurator, certificate.Manager) (*xds_discovery.DiscoveryResponse, error),
+	meshCatalog catalog.MeshCataloger, proxy *envoy.Proxy, request *xds_discovery.DiscoveryRequest, cfg configurator.Configurator, certIssuer certificate.Manager,
+) (*xds_discovery.DiscoveryResponse, error) {
+	log.Debug().Msgf("Looking for %s Response in cache first", dsType)
+
+	if memoizationCache[dsType] == nil {
+		log.Debug().Msgf("Initializing %s memoization cache", dsType)
+		memoizationCache[dsType] = cache.New(defaultExpiration, cleanupInterval)
```

Comment:
-- for discussion only --
I'm assuming you are just timing how much do you live things in the cache with the expiration parametres.

IMHO This will very hardly work on the long run. Being time based, in some situations this will work very well, in some others this will not work at all.
I can think many cases things will break, or you will need to bring back tickers and resyncs because caches will be inconsistent with current state and will return you old values. It's also going to be very non-deterministic as with one config, one function will return different results depending on timing.

For these black-box-like xds caching model to work, IMO you'd need to be able to understand when to evict the caches consistently, and not base it on time.